src/library/scala/collection/Iterable.scala

```diff
@@ -912,3 +905,109 @@ object Iterable extends IterableFactory.Delegate[Iterable](immutable.Iterable) {
+  * the `CC` and `C` type parameters.
+  *
+  * The default implementations in this trait can be used in the common case when `CC[A]` is the
+  * same as `C`.
+  */
+trait IterableFactoryDefaults[A, +CC[x] <: IterableOps[x, CC, CC[x]]] extends IterableOps[A, CC, CC[A]] {
+  protected def fromSpecific(coll: IterableOnce[A]): CC[A] = iterableFactory.from(coll)
+  protected def newSpecificBuilder: Builder[A, CC[A]] = iterableFactory.newBuilder[A]
+
+  // overridden for efficiency, since we know CC[A] =:= C
+  override def empty: CC[A] = iterableFactory.empty
+}
+
+/** This trait provides default implementations for the factory methods `fromSpecific` and
+  * `newSpecificBuilder` that need to be refined when implementing a collection type that refines
+  * the `CC` and `C` type parameters. It is used for collections that have an additional constraint,
+  * expressed by the `evidenceIterableFactory` method.
+  *
+  * The default implementations in this trait can be used in the common case when `CC[A]` is the
+  * same as `C`.
+  */
+trait EvidenceIterableFactoryDefaults[A, +CC[x] <: IterableOps[x, CC, CC[x]], Ev[_]] extends IterableOps[A, CC, CC[A]] {
+  protected def evidenceIterableFactory: EvidenceIterableFactory[CC, Ev]
+  implicit protected def iterableEvidence: Ev[A]
+  override protected def fromSpecific(coll: IterableOnce[A]): CC[A] = evidenceIterableFactory.from(coll)
+  override protected def newSpecificBuilder: Builder[A, CC[A]] = evidenceIterableFactory.newBuilder[A]
+  override def empty: CC[A] = evidenceIterableFactory.empty
+}
+
+/** This trait provides default implementations for the factory methods `fromSpecific` and
+  * `newSpecificBuilder` that need to be refined when implementing a collection type that refines
+  * the `CC` and `C` type parameters. It is used for sorted sets.
+  *
+  * Note that in sorted sets, the `CC` type of the set is not the same as the `CC` type for the
+  * underlying iterable (which is fixed to `Set` in [[SortedSetOps]]). This trait has therefore
+  * two type parameters `CC` and `WithFilterCC`. The `withFilter` method inherited from
+  * `IterableOps` is overridden with a compatible default implementation.
+  *
+  * The default implementations in this trait can be used in the common case when `CC[A]` is the
+  * same as `C`.
+  */
+trait SortedSetFactoryDefaults[A,
+    +CC[X] <: SortedSet[X] with SortedSetOps[X, CC, CC[X]],
+    +WithFilterCC[x] <: IterableOps[x, WithFilterCC, WithFilterCC[x]] with Set[x]] extends SortedSetOps[A, CC, CC[A]] {
+  self: IterableOps[A, WithFilterCC, _] =>
+
+  override protected def fromSpecific(coll: IterableOnce[A]): CC[A]    = sortedIterableFactory.from(coll)
+  override protected def newSpecificBuilder: mutable.Builder[A, CC[A]] = sortedIterableFactory.newBuilder[A]
+  override def empty: CC[A] = sortedIterableFactory.empty
+
+  override def withFilter(p: A => Boolean): SortedSetOps.WithFilter[A, WithFilterCC, CC] =
```

Comment:
Having `withFilter` in these traits seems odd, but then again, `withFilter` was always the ugliest and most problematic part of the new collections design.